internal/dedupe/redis_dedupe/redis_dedupe.go

```diff
@@ -0,0 +1,55 @@
+package redis_dedupe
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/event"
+	"github.com/redis/go-redis/v9"
+	"golang.org/x/exp/slog"
+)
+
+type Dedupe struct {
+	config DedupeConfig
+}
+
+type DedupeConfig struct {
+	Logger     *slog.Logger
+	Redis      *redis.Client
+	Expiration time.Duration
+}
+
+func NewDedupe(config DedupeConfig) Dedupe {
+	if config.Expiration == 0 {
+		config.Expiration = 24 * time.Hour
+	}
+
+	return Dedupe{
+		config,
+	}
+}
+
+// IsUnique checks if the entry is unique based on the key
+func (d *Dedupe) IsUnique(ctx context.Context, ev event.Event) (bool, error) {
+	isSet, err := d.config.Redis.Exists(ctx, getKey(ev)).Result()
+	if err != nil {
+		return false, err
+	}
+	return isSet == 0, nil
+}
+
+// Set sets the entry in the dedupe store
+func (d *Dedupe) Set(ctx context.Context, ev event.Event) (bool, error) {
+	isUnique, err := d.config.Redis.SetNX(ctx, getKey(ev), "", d.config.Expiration).Result()
+	if err != nil {
+		return isUnique, err
+	}
+
+	return isUnique, nil
+}
+
+// CloudEvents are unique based on the source and id
+func getKey(ev event.Event) string {
+	return fmt.Sprintf("%s-%s", ev.Source(), ev.ID())
```

Comment:
This is problematic in the sense that ID's need to be globally unique, but there is nothing to enforce that at the moment. An ID submitted by one tenant would transparently be deduplicated. We need to add some tenancy information in here (namespace?) to make sure deduplication only happens below the tenant level.